### PR TITLE
Various diagnostics, particularly about SMT queries

### PR DESCRIPTION
**Pull request by nikswamy**
Given a file like this with a couple of reset-options

```
module Test

let arr = int -> Tot int
let ref_arr = f:arr{forall x. f x == x}

assume val f : arr
assume val g : int -> Tot int
assume val h : ref_arr
assume val i : unit -> ref_arr

#reset-options "--query_stats"
let test (x:int) =
  assert (i () x == x)

let test2 (x:int) =
  assert (i () (x + 1) == (1 + x))

#reset-options "--query_stats --using_facts_from 'Prims Test'"
let test3 (x:int) =
  assert (i () (x + 2) == (2 + x))
```

running

```
$ fstar test.fst |& grep "^Query Stats:"
Query Stats: context_profile with 2911 assertions
Query Stats: module FStar.Pervasives.Native produced 1321 SMT decls
Query Stats: module FStar.Pervasives produced 944 SMT decls
Query Stats: module Prims produced 646 SMT decls
Query Stats: (test.fst(13,2-13,22))      (Test.test, 1) succeeded in 550 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats:Specifically:
Query Stats:    Prims
Query Stats:    Test.arr
Query Stats:    Test.ref_arr
Query Stats:    Prims.__cache_version_number__
Query Stats:    Test
Query Stats:    Test.i
Query Stats:    Test.i
Query Stats: (test.fst(16,2-16,34))      (Test.test2, 1)        succeeded in 44 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats:Specifically:
Query Stats:    Prims
Query Stats:    Test.arr
Query Stats:    Test.ref_arr
Query Stats:    Prims.__cache_version_number__
Query Stats:    Prims.op_Addition
Query Stats:    Test
Query Stats:    Test.i
Query Stats:    Test.i
Query Stats: context_profile with 646 assertions
Query Stats: module Prims produced 646 SMT decls
Query Stats: (test.fst(20,2-20,34))      (Test.test3, 1)        succeeded in 99 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats:Specifically:
Query Stats:    Prims
Query Stats:    Test.arr
Query Stats:    Test.ref_arr
Query Stats:    Prims.__cache_version_number__
Query Stats:    Prims.op_Addition
Query Stats:    Test
Query Stats:    Test.i
Query Stats:    Test.i
```

When `--query_stats` is true:

At each #reset-options you get a context profile, telling you how many top-level SMT facts each module contributed.

Further, for each query, we report back from the unsat core which modules, and specifically which definitions in those modules, were used.

I'd like to see more stuff surfacing through this interface. 

Notably stuff from smt.qi.profile (@wintersteiger)

But, feedback on what there is now is most welcome.

### CODE REVIEW

**Comment by wintersteiger** [comment]
CC @arpj-rebola

File: src/smtencoding/FStar.SMTEncoding.Solver.fs [hunk 1]
```diff
@@ -310,14 +311,95 @@ let report_errors settings : unit =
...
+       in
+       let add_discarded_name, get_discarded_names =
+           accumulator()
+       in
+       let parse_axiom_name (s:string) =
+            let chars = String.list_of_string s in
+            let first_upper_index =
+                BU.try_find_index BU.is_upper chars
+            in
+            match first_upper_index with
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.Solver.fs]: Can you please add a comment what is trying to filter out?

File: src/basic/FStar.Options.fs [hunk 2]
```diff
@@ -1431,6 +1438,14 @@ let hint_file                    () = get_hint_file                   ()
 let ide                          () = get_ide                         ()
 let print                        () = get_print                       ()
 let print_in_place               () = get_print_in_place              ()
+let profile (f:unit -> 'a) (msg:'a -> string) : 'a =
```
Comment (aseemr) [inline src/basic/FStar.Options.fs]: We could print the profiling message like `Starting task <string> ... Ending task <string>, total time taken: ...`. That way the nesting of  `Options.profile` calls will be clearer in the output.

But it is minor, we can do this later as we try to use `Options.profile` more.

**Comment by nikswamy** [comment]
Thanks for the review. Here's what it looks like now:


```
$ fstar test.fst |& grep "^Query Stats:"
Query Stats: context_profile with 2911 assertions
Query Stats: module FStar.Pervasives.Native produced 1321 SMT decls
Query Stats: module FStar.Pervasives produced 944 SMT decls
Query Stats: module Prims produced 646 SMT decls
Query Stats: (test.fst(13,2-13,22))      (Test.test, 1) succeeded in 492 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats: (test.fst(16,2-16,34))      (Test.test2, 1)        succeeded in 43 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats: context_profile with 646 assertions
Query Stats: module Prims produced 646 SMT decls
Query Stats: (test.fst(20,2-20,34))      (Test.test3, 1)        succeeded in 100 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test

nswamy@nikhil-Z820 /cygdrive/c/Users/nswamy/workspace/test2
$ fstar test.fst |& grep "^Query Stats"
Query Stats: context_profile with 2911 assertions
Query Stats: module FStar.Pervasives.Native produced 1321 SMT decls
Query Stats: module FStar.Pervasives produced 944 SMT decls
Query Stats: module Prims produced 646 SMT decls
Query Stats: (test.fst(13,2-13,22))      (Test.test, 1) succeeded in 493 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats (Detail Z3): statistics={mk-bool-var=6986 del-clause=1454 num-checks=1 conflicts=1 binary-propagations=2918 propagations=3347 rlimit-count=357627 quant-instantiations=416 mk-clause=1533 memory=16.54 max-generation=5 time=0.04 num-allocs=45407730 datatype-accessor-ax=3 max-memory=16.76 added-eqs=516}
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats (Detail 1): Specifically:
Query Stats (Detail 1): Prims
Query Stats (Detail 1): Test.arr
Query Stats (Detail 1): Test.ref_arr
Query Stats (Detail 1): Prims.__cache_version_number__
Query Stats (Detail 1): Test
Query Stats (Detail 1): Test.i
Query Stats (Detail 1): Test.i
Query Stats (Detail 2): Note, this report ignored the following names in the context: @MaxIFuel_assumption, @query, int_inversion, unit_typing
Query Stats: (test.fst(16,2-16,34))      (Test.test2, 1)        succeeded in 43 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats (Detail Z3): statistics={mk-bool-var=8132 del-clause=2938 num-checks=2 conflicts=2 binary-propagations=5836 propagations=6705 arith-assert-upper=2 arith-assert-lower=2 rlimit-count=377663 arith-offset-eqs=1 quant-instantiations=851 mk-clause=3021 memory=16.63 arith-pivots=2 max-generation=5 time=0.04 num-allocs=48090378 datatype-accessor-ax=5 max-memory=16.86 arith-eq-adapter=2 added-eqs=1051}
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats (Detail 1): Specifically:
Query Stats (Detail 1): Prims
Query Stats (Detail 1): Test.arr
Query Stats (Detail 1): Test.ref_arr
Query Stats (Detail 1): Prims.__cache_version_number__
Query Stats (Detail 1): Prims.op_Addition
Query Stats (Detail 1): Test
Query Stats (Detail 1): Test.i
Query Stats (Detail 1): Test.i
Query Stats (Detail 2): Note, this report ignored the following names in the context: @MaxIFuel_assumption, @query, int_typing, unit_typing
Query Stats: context_profile with 646 assertions
Query Stats: module Prims produced 646 SMT decls
Query Stats: (test.fst(20,2-20,34))      (Test.test3, 1)        succeeded in 104 milliseconds with fuel 2 and ifuel 1 and rlimit 2723280
Query Stats (Detail Z3): statistics={mk-bool-var=1865 del-clause=651 num-checks=1 conflicts=1 binary-propagations=648 propagations=870 arith-assert-upper=2 arith-assert-lower=2 rlimit-count=70511 arith-offset-eqs=1 quant-instantiations=176 mk-clause=696 memory=6.05 arith-pivots=2 max-generation=5 time=0.01 num-allocs=2983846 datatype-accessor-ax=3 max-memory=6.05 arith-eq-adapter=2 added-eqs=212}
Query Stats: Modules relevant to this proof:
Query Stats:    Prims
Query Stats:    Test
Query Stats (Detail 1): Specifically:
Query Stats (Detail 1): Prims
Query Stats (Detail 1): Test.arr
Query Stats (Detail 1): Test.ref_arr
Query Stats (Detail 1): Prims.__cache_version_number__
Query Stats (Detail 1): Prims.op_Addition
Query Stats (Detail 1): Test
Query Stats (Detail 1): Test.i
Query Stats (Detail 1): Test.i
Query Stats (Detail 2): Note, this report ignored the following names in the context: @MaxIFuel_assumption, @query, int_typing, unit_typing
```

**Comment by aseemr** [comment]
@nikswamy, @wintersteiger, this broke the hint failures reporting on CI, since that script parses specific strings in the output log. The log size also became huge since `--query-stats` is enabled in CI by default.

So I changed it so that additional details are printed only when `--print_z3_statistics` is on. For `--query_stats` the behavior is as before. Also, the prefix string in new messages is "Z3 Proof Stats".

**Pull request merged**